### PR TITLE
Update Docs + New Generate Did Endpoint

### PR DESCRIPTION
**Pull request by madaster97**
# Changes in prep for sidetree updates:
1. Added a testing-endpoint `api/test/generate-did` that can take in public keys in JWK format and create a Long Form DID for that entity
2. Updated the `generateDid` export of `src/dids.ts` to do the following:
   - Accept multiple `domain` strings (expected an array, but only used first entry before)
   - Make the `signing` and `encryption` public JWKs **optional**, since not all entities will need to both sign and encrypt data
3. Update testing-endpoints documentation:
   - Use re-formated DIDs in examples (JWE and JWS payloads TODO noted)
   - Add documentation for the `generate-did` API

### CODE REVIEW

File: testing-endpoints.md [hunk 1]
```diff
@@ -18,20 +18,60 @@ curl -X GET http://localhost:8080/api/test/did-debug
...
+Returns a Long-Form did made from input public keys in jwk format
+
+The body is a JSON object (note, this API requires a content-type header) with the following fields:
+1. `recoveryPublicJwk` - **required** `ES256` public signing key jwk for did recovery operations
+2. `updatePublicJwk` - **required** `ES256` public signing key jwk for did update operations
+3. `signingPublicJwk` - **optional** `ES256` public signing key jwk for verifying assertions 
+4. `encryptionPublicJwk` - **optional** public encryption key jwk using the `"alg": "ECDH-ES"` and `"enc": "A256GCM"` encryption algorithm
+5. `domains` - **optional** array of string URI hostnames. Will be added to the did's `LinkedDomains` by an `add-services` action. 
+
+**Note**: The `signingPublicJwk` and `encryptionPublicJwk` fields are optional since
```
Comment (dleve123) [inline testing-endpoints.md:32]: While it's technically true that sidetree doesn't require a DID Document to include signing and encryption keys, the health cards spec does:

> The response is signed as a JWS with the user's DID and optionally encrypted using the lab's DID (if the request specified id_token_encrypted_response_*). The latter step requires looking inside the DID Document for an encryption key, which can be used for encrypting a payload for this party.

(https://healthwallet.cards/#connect-health-wallet-to-lab-account)

As such, my 2¢ is to require the signing key to be present -- just to facilitate those getting started with health cards to have as few roadblocks as possible.

File: testing-endpoints.md [hunk 2]
```diff
@@ -18,20 +18,60 @@ curl -X GET http://localhost:8080/api/test/did-debug
...
+1. `recoveryPublicJwk` - **required** `ES256` public signing key jwk for did recovery operations
+2. `updatePublicJwk` - **required** `ES256` public signing key jwk for did update operations
+3. `signingPublicJwk` - **optional** `ES256` public signing key jwk for verifying assertions 
+4. `encryptionPublicJwk` - **optional** public encryption key jwk using the `"alg": "ECDH-ES"` and `"enc": "A256GCM"` encryption algorithm
+5. `domains` - **optional** array of string URI hostnames. Will be added to the did's `LinkedDomains` by an `add-services` action. 
+
+**Note**: The `signingPublicJwk` and `encryptionPublicJwk` fields are optional since 
+not all entities represented by DIDs may need to sign or encrypt documents.
+
+For exmample, a `verifier` or `issuer` entity that does not request SIOP Responses to be encrypted MAY only need to have a signing key (for signing SIOP requests), and not an encryption key.
```
Comment (dleve123) [inline testing-endpoints.md:35]:
```suggestion
For example, a `verifier` or `issuer` entity that does not request SIOP Responses to be encrypted MAY only need to have a signing key (for signing SIOP requests), and not an encryption key.
```

However, I think this doc line even further suggests that we should automatically create signing keys.

File: testing-endpoints.md [hunk 3]
```diff
@@ -18,20 +18,60 @@ curl -X GET http://localhost:8080/api/test/did-debug
...
+
+Example key props for signing keys:
+```
+{
+    "kty": "EC",
+    "crv": "P-256",
+    "alg": 'ES256',
```
Comment (dleve123) [inline testing-endpoints.md:42]: include `"x":"rAkPWQoZai50iSgCkBQ-eHR9O4L7edymIAUB1m9O6c8","y":"H-ulzimHfP-LyJYpPeIUoNguNs7jFfqaCncE6ZKkLUc"` for completeness? Similar for encryption key prop below?

**Review by dleve123 — COMMENTED** [review]
Overall this PR is looking good to me! Per my inline comment, I think it would be best to require the signing key from a dev UX perspective, but that's just my opinion.

File: testing-endpoints.md (hunk 1 above)
Comment (madaster97) [reply testing-endpoints.md:32]: I'm on board with this. I'll work on making this required in 1) the documentation, 2) the API, 3) the internal function called

File: testing-endpoints.md (hunk 3 above)
Comment (madaster97) [reply testing-endpoints.md:42]: I think I'd rather just remove these objects. I only put them in as reference to the JOSE library inputs, but I'd rather omit than try to be too rigorous here.

Comment (madaster97) [reply testing-endpoints.md:42]: Also, there are example JWKs in the request, and those have all the values!

**Comment by jmandel** [comment]
This is awesome - sorry again that I missed the GH notifications. The only thing I'm unclear about is the "custom suffix" option -- why is this needed, and would it be possible to get by without injecting test-related functionality into the main did generation code?

**Comment by madaster97** [comment]
I removed the custom suffix. I realize that people can just splice in whatever suffix they want, so no reason to have it!

File: src/dids.ts [hunk 4]
```diff
@@ -48,32 +48,47 @@ export async function generateDid({ signingPublicJwk, encryptionPublicJwk, recov
...
-                "type": "LinkedDomains",
-                "serviceEndpoint": domains[0]
-                }]
-            });
-    } 
+            "services": domains.map((domain, index) => {
+                return {
+                    "id": `linked-domain-${index + 1}`,
+                    "type": "LinkedDomains",
+                    "serviceEndpoint": domain
```
Comment (jmandel) [inline src/dids.ts:87]: Technically there's an option to provide an array of origins within a single `LinkedDomains` entry in the `services[]` array, but I think this "exploded" list is cleaner / more likely to be compatible with client implementations (and ... we only have a use case for one right now anyway)

File: testing-endpoints.md [hunk 5]
```diff
@@ -18,20 +18,43 @@ curl -X GET http://localhost:8080/api/test/did-debug
...
+
+The body is a JSON object (note, this API requires a content-type header) with the following fields:
+1. `recoveryPublicJwk` - **required** `ES256` public signing key jwk for did recovery operations
+2. `updatePublicJwk` - **required** `ES256` public signing key jwk for did update operations
+3. `signingPublicJwk` - **required** `ES256` public signing key jwk for verifying assertions 
+4. `encryptionPublicJwk` - **optional** public encryption key jwk using the `"alg": "ECDH-ES"` and `"enc": "A256GCM"` encryption algorithm
+5. `domains` - **optional** array of string URI hostnames. Will be added to the did's `LinkedDomains` by an `add-services` action. 
+
+**Note**: The `encryptionPublicJwk` field is optional since 
+not all entities represented by DIDs may need to encrypt documents.
```
Comment (jmandel) [inline testing-endpoints.md:33]: This key is technically used to *received* encrypted documents; I'm happy leaving it optional here though.

File: testing-endpoints.md [hunk 6]
```diff
@@ -18,20 +18,43 @@ curl -X GET http://localhost:8080/api/test/did-debug
...
+
+For example, a `verifier` or `issuer` entity that does not request SIOP Responses to be encrypted MAY only need to have a signing key (for signing SIOP requests), and not an encryption key.
```
Comment (jmandel) [inline testing-endpoints.md:35]: Yes, this clarifies.

**Comment by jmandel** [comment]
@madaster97 Thanks for cleaning up that param. Are you willing to rebase this against the current master (maybe squash into one commit)? (Otherwise I can take care of it, but may take me a little while to get there.)

**Comment by madaster97** [comment]
Merged! I just did a diff against master and everything looks good.

I was having an issue with dependencies, but fixed those in my most recent commit. I did a test and was able to spin up the server and test the endpoint.

**Comment by madaster97** [comment]
@jmandel I can open another PR against master if that's easier (or dev if you'd like to spin up a general dev branch again).

It looks like a ton of changes for this PR, but diffing my branch against master shows only a handful of changes.

**Comment by jmandel** [comment]
Thanks so much. if you don't mind opening a new pull request against `master` that consolidates these commits and directly reflects the handful of actual changes that would be much appreciated.

**Comment by madaster97** [comment]
Did who? Closing this PR.

**Pull request closed without merging**